src/core/exec-invoke.c

```diff
@@ -2251,6 +2251,44 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
         return 1;
 }
 
+static int setup_private_pids(const ExecContext *c, ExecParameters *p) {
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+        int r, q;
+
+        if (unshare(CLONE_NEWPID) < 0)
+                return log_exec_debug_errno(c, p, errno, "Failed to unshare pid namespace: %m");
+
+        /* The first process created after unsharing a pid namespace becomes PID 1 in the pid namespace, so
+         * we have to fork after unsharing the pid namespace to become PID 1. The parent sends the child
+         * pidref to the manager and exits while the child process continues with the rest of exec_invoke()
+         * and finally executes the actual payload. */
+
+        r = pidref_safe_fork("(sd-pidfd-child)", /*flags=*/ 0, &pidref);
+        if (r < 0)
+                return log_exec_debug_errno(c, p, r, "Failed to fork child into new pid namespace: %m");
+
+        if (r > 0) {
+                /* In the parent process, we send the child pidref to the manager and exit. */
+
+                if (p->pidref_transport_fd >= 0) {
+                        q = send_one_fd_iov(
+                                        p->pidref_transport_fd,
+                                        pidref.fd,
+                                        &IOVEC_MAKE(&pidref.pid, sizeof(pidref.pid)),
+                                        /*iovlen=*/ 1,
+                                        /*flags=*/ 0);
+                        if (q < 0)
+                                return log_exec_debug_errno(c, p, q, "Failed to send child pidref to manager: %m");
+                }
+
+                /* Our caller will make sure the "middle" process will exit ASAP. */
+        }
+
+        p->pidref_transport_fd = safe_close(p->pidref_transport_fd);
+
```

Comment:
Can you add a big fat comment here, that highlights the fact that this function returns twice. that is an absolute exception in our codebase.

i.e. add here:

```c
/* NOTE! This function returns TWICE, once in the parent and once in the child, and it's the *child* that is going to survive, while the parent will exit shortly! */
```